pkg/upstream/cluster/strict_dns_cluster.go

```diff
@@ -0,0 +1,311 @@
+	"mosn.io/pkg/utils"
+)
+
+func init() {
+	RegisterClusterType(v2.STRICT_DNS_CLUSTER, newStrictDnsCluster)
+}
+
+type strictDnsCluster struct {
+	*simpleCluster
+	dnsResolver     *network.DnsResolver
+	dnsLookupFamily v2.DnsLookupFamily
+	respectDnsTTL   bool
+	resolveTargets  []*ResolveTarget
+	dnsRefreshRate  time.Duration
+	mutex           sync.Mutex
+}
+
+var DefaultRefreshTimeout time.Duration = 20 * time.Second
+var DefaultRefreshInterval time.Duration = 5 * time.Second
+
+type ResolveTarget struct {
+	config           *v2.Host
+	dnsAddress       string
+	port             string
+	strictDnsCluster *strictDnsCluster
+	resolveTimer     *utils.Timer
+	resolveTimeout   *utils.Timer
+	refreshTimeout   time.Duration
+	dnsRefreshRate   chan time.Duration
+	stop             chan struct{}
+	timeout          chan bool
+	resolveCount     uint32
+	hosts            []types.Host
+}
+
+func newStrictDnsCluster(clusterConfig v2.Cluster) types.Cluster {
+
+	cluster := &strictDnsCluster{
+		simpleCluster:   newSimpleCluster(clusterConfig).(*simpleCluster),
+		dnsLookupFamily: clusterConfig.DnsLookupFamily,
+		respectDnsTTL:   clusterConfig.RespectDnsTTL,
+		resolveTargets:  []*ResolveTarget{},
+		mutex:           sync.Mutex{},
+	}
+
+	// set dnsRefreshRate
+	if clusterConfig.DnsRefreshRate != nil {
+		cluster.dnsRefreshRate = clusterConfig.DnsRefreshRate.Duration
+	}
+
+	// set resolve server
+	if clusterConfig.DnsResolverConfig.Servers != nil {
+		cluster.dnsResolver = network.NewDnsResolver(&clusterConfig.DnsResolverConfig)
+	} else {
+		cluster.dnsResolver = network.NewDnsResolverFromFile(clusterConfig.DnsResolverFile, clusterConfig.DnsResolverPort)
+	}
+
+	return cluster
+}
+
+// supported formats including {aaa.com:80, aaa.com}
+func getHostPortFromAddr(addr string) (string, string) {
+	s := strings.Split(addr, ":")
+	if len(s) == 1 {
+		return s[0], ""
+	} else {
+		return s[0], s[1]
+	}
+}
+
+func (sdc *strictDnsCluster) UpdateHosts(newHosts []types.Host) {
+	sdc.mutex.Lock()
+	defer sdc.mutex.Unlock()
+	sdc.StopResolve()
+
+	// before resolve, init hosts to origin unresolved address
+	sdc.simpleCluster.UpdateHosts(newHosts)
+
+	// update resolve targets and start new resolve task
+	var rts []*ResolveTarget
+	for _, host := range newHosts {
+		addr, port := getHostPortFromAddr(host.AddressString())
+		if addr == "" {
+			if log.DefaultLogger.GetLogLevel() >= log.ERROR {
+				log.DefaultLogger.Errorf("[upstream] [strict_dns_cluster] config address format error: %s", host.AddressString())
+			}
+			continue
+		}
+		// default port: 80
+		if port == "" {
+			port = "80"
+		}
+		config := host.Config()
+		rt := &ResolveTarget{
+			dnsAddress:       addr,
+			port:             port,
+			config:           &config,
+			refreshTimeout:   DefaultRefreshTimeout,
+			dnsRefreshRate:   make(chan time.Duration),
+			stop:             make(chan struct{}),
+			timeout:          make(chan bool),
+			strictDnsCluster: sdc,
+			hosts:            []types.Host{host},
```

Comment:
Maybe we need to add a `version` field to `ResolveTarget` struct to determine whether it is the same as the cluster version. If they are inconsistent, there is no need to call `sdc.simpleCluster.UpdateHosts (newHosts)` in `updateDynamicHosts`. forexample：

```
func (sdc *strictDnsCluster) UpdateHosts(newHosts []types.Host) {  
    ……
    atomic.AddUint64(&sdc.version , 1)
   ……
}

func (sdc *strictDnsCluster) updateDynamicHosts(newHosts []types.Host, rt *ResolveTarget) {
……
v := atomic.LoadUint64(&sdc.version)
// don't overwriting 
if rt.version != v {
    continue
}
……

}

```